worker/rhpv2.go

```diff
@@ -261,3 +261,106 @@ func (w *worker) FetchSignedRevision(ctx context.Context, hostIP string, hostKey
 	})
 	return rev, err
 }
+
+func (w *worker) FetchContractRoots(ctx context.Context, hostIP string, hostKey types.PublicKey, renterKey types.PrivateKey, contractID types.FileContractID, timeout time.Duration) (roots []types.Hash256, err error) {
+	err = w.withTransportV2(ctx, hostKey, hostIP, func(t *rhpv2.Transport) error {
+		req := &rhpv2.RPCLockRequest{
+			ContractID: contractID,
+			Signature:  t.SignChallenge(renterKey),
+			Timeout:    uint64(time.Minute.Milliseconds()),
+		}
+
+		// execute lock RPC
+		var lockResp rhpv2.RPCLockResponse
+		if err := t.Call(rhpv2.RPCLockID, req, &lockResp); err != nil {
+			return err
+		}
+		t.SetChallenge(lockResp.NewChallenge)
+
+		// extract the revision
+		rev := rhpv2.ContractRevision{
+			Revision:   lockResp.Revision,
+			Signatures: [2]types.TransactionSignature{lockResp.Signatures[0], lockResp.Signatures[1]},
+		}
+
+		// defer unlock RPC
+		defer t.WriteRequest(rhpv2.RPCUnlockID, nil)
+
+		// execute settings RPC
+		var settingsResp rhpv2.RPCSettingsResponse
+		if err := t.Call(rhpv2.RPCSettingsID, nil, &settingsResp); err != nil {
+			return err
+		}
+		var settings rhpv2.HostSettings
+		if err := json.Unmarshal(settingsResp.Settings, &settings); err != nil {
+			return fmt.Errorf("couldn't unmarshal json: %w", err)
+		}
+
+		// download all roots
```

Comment:
I think there are two things missing here.

1. I'd add a sanity check to compare the returned revision number to our own latest known one to avoid having the host slip us an outdated one
2. verify not just the host sig but also our own on the revision to make sure it's a revision we have signed off on in the past

This is an excellent way for a host to try and get us to delete sectors that they lost and we should try and minimise any chance of that happening.